packages/DataGrid/src/DataGrid.js

```diff
@@ -0,0 +1,628 @@
+
+const innerElementTypeMainGrid = React.forwardRef((props, ref) => (
+  <styled.InnerElementTypeMainGrid className="inner-element-type-main-grid" role="row" ref={ref} {...props} />
+));
+
+const DataGrid = React.forwardRef((props, ref) => {
+  const {
+    children,
+    data,
+    hasNoRecords,
+    height,
+    isIdle,
+    onClick,
+    onEnter,
+    onKeyDown,
+    onRowChecked,
+    onShiftSpaceBar,
+    onSpaceBar,
+    rowHeight,
+    width,
+    ...moreProps
+  } = props;
+
+  const refsCell = React.useRef({ keys: {}, rows: {} });
+  const refPrevActiveCell = React.useRef(null);
+  const refScrollHeader = React.useRef(null);
+  const refContainer = React.useRef(null);
+  const refGrid = React.useRef(null);
+  const refScrollStickyColumns = React.useRef(null);
+  const refScrollGrid = React.useRef(null);
+  const refScrollHappenedBy = React.useRef(null);
+  const refEnd = React.useRef(null);
+  const refPrevLastScrollHeight = React.useRef(null);
+  const refCurrentPage = React.useRef(null);
+  const refActiveRow = React.useRef(null);
+  const refPrevActiveRow = React.useRef(null);
+  const refVisibleIndexes = React.useRef({
+    start: null,
+    stop: null,
+  });
+
+  const [scrollBarWidth, setScrollBarWidth] = React.useState(0);
+  const [gridShouldHaveFocus, setGridShouldHaveFocus] = React.useState(true);
+  const [pageSize, setPageSize] = React.useState(null);
+  // these two value are sensitive in Grids with lots of columns and can degradate performance alot.
+  // be caution when using them.
+  const overscanRowCount = 5;
+  const overscanColumnCount = 2;
+
+  const rowCount = React.useMemo(() => {
+    return data.length;
+  }, [data.length]);
+
+  const { ColumnDefinitions, WhenScrollBarReachedBottom, InfinityScroll } = React.useMemo(() => {
+    const {
+      "DataGrid.ColumnDefinition": ColumnDefinitions,
+      "DataGrid.WhenScrollBarReachedBottom": WhenScrollBarReachedBottom,
+      "DataGrid.InfinityScroll": InfinityScroll,
+    } = extractChildren(children, [
+      "DataGrid.ColumnDefinition",
+      "DataGrid.WhenScrollBarReachedBottom",
+      "DataGrid.InfinityScroll",
+    ]);
+
+    return { ColumnDefinitions, WhenScrollBarReachedBottom, InfinityScroll };
+  }, [children]);
+
+  const columnCount = React.useMemo(() => {
+    return ColumnDefinitions.length;
+  }, [ColumnDefinitions]);
+
+  const columnHeadersA11yText = React.useMemo(() => {
+    return ColumnDefinitions.map(ColumnDefinition => {
+      const { header, headerA11yText } = ColumnDefinition.props;
+      return typeof header === "function" ? headerA11yText && headerA11yText() : header;
+    });
+  }, [ColumnDefinitions]);
+
+  const calculatedTableHeight = React.useMemo(() => {
+    return rowHeight * rowCount > height ? height : rowHeight * rowCount + scrollBarWidth;
+  }, [height, rowCount, rowHeight, scrollBarWidth]);
+
+  const calculatedTableWidth = React.useCallback(() => {
+    let width = 0;
+    ColumnDefinitions.forEach(ColumnDefinition => {
```

Comment:
Maybe`.reduce`? If so, you could also do it in `stickyGridWidth`

Reply:
I actually implemented this with a reducer but I think is way simpler to read with forEach. https://youtu.be/qaGjS7-qWzg